nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,313 @@
+    # prioritize small messages coming from the writer (user
+    # creations/updates/deletes).
+    #
+    # Priority order: lower values first
+    priorities = {
+        IndexMessageSource.WRITER: 0,
+        IndexMessageSource.PROCESSOR: 1,
+    }
+
+    index_message: IndexMessage
+    nats_msg: Msg
+    mpu: MessageProgressUpdater
+
+    def __init__(self, *args, **kwargs):
+        raise Exception("__init__ method not allowed. Use a from_* method instead")
+
+    @classmethod
+    def __new__(
+        cls, *, index_messsage: IndexMessage, nats_msg: Msg, mpu: MessageProgressUpdater
+    ):
+        instance = super().__new__(cls)
+        instance.index_message = index_messsage
+        instance.nats_msg = nats_msg
+        instance.mpu = mpu
+        return instance
+
+    @classmethod
+    def from_msg(cls, msg: Msg) -> "WorkUnit":
+        mpu = MessageProgressUpdater(msg, nats_consumer_settings.nats_ack_wait * 0.66)
+
+        pb = IndexMessage()
+        pb.ParseFromString(msg.data)
+
+        return cls.__new__(index_messsage=pb, nats_msg=msg, mpu=mpu)
+
+    @property
+    def seqid(self) -> int:
+        return int(self.nats_msg.reply.split(".")[5])
+
+    @property
+    def shard_id(self) -> str:
+        return self.index_message.shard
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self._priority_id.__eq__(other._priority_id)
+
+    @cached_property
+    def _priority_id(self) -> tuple[int, int]:
+        # Priority based on message source and smaller seqid
+        source_priority = self.priorities[self.index_message.source]
+        return (source_priority, self.seqid)
+
+    def __lt__(self, other):
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self._priority_id.__lt__(other._priority_id)
+
+
+class ConcurrentShardIndexer:
+    def __init__(self, writer: Writer):
+        self.writer = writer
+        self.workers: dict[str, PriorityIndexer] = {}
+
+    async def initialize(self):
+        self.storage = await get_storage(service_name=SERVICE_NAME)
+
+    async def finalize(self):
+        for indexer in self.workers.values():
+            await indexer.finalize()
+        self.workers.clear()
+
+    async def index_message_soon(self, msg: Msg):
+        work = WorkUnit.from_msg(msg)
+        await work.mpu.start()
+
+        logger.info(
+            f"Index message for shard {work.shard_id} is being enqueued",
+            extra={
+                "shard": work.shard_id,
+                "subject": msg.subject,
+                "reply": msg.reply,
+                "seqid": work.seqid,
+                "storage_key": work.index_message.storage_key,
+            },
+        )
+
+        indexer, created = await self.get_or_create_indexer(work.shard_id)
+        await indexer.index_soon(work)
+        await self.clean_idle_indexers()
```

Comment:
nit: should this be a background task managed by the class instance instead? I'm fine as it is now too, but we will only clean indexers if new messages arrive